### DISCUSSION TITLE
EV/PHEV Battery Pack Monitoring

### DISCUSSION THREAD

**Discussion by smurf12345** [meatpiHQ/wican-fw#23] [General]
Hi All:
I just received my WiCAN and hope to be able to pull battery back information along with other data points.  So far I was able to pull the mode 1 battery pack remaining information using the examples in the documentation.  I'm using Node-Red to collect this information in Home Assistant.

I've set the WiCAN sleep mode to go to sleep when voltage is below 13volts.  This way as soon as the car is put in run/start, the WiCAN wakes up and starts pulling information.   I was hoping the ECU would wake up while plugged in and active charging was happing when the car was off, but looks like the ECU stays asleep.

I would love to see some people's examples of pulling Mode 22 battery information if it is even possible.  I have found some information out there on my car (Pacifica Hybrid Plugin) specific PIDs, but haven't made sense of it yet.

Here is what I have so far using Mode 1.  I maybe getting multiple answers back from different modules with all the same information, so I may need to filter the receive message a bit more.

Hybrid Battery Pack Remaining:

Transmit:
```
msg.payload = { "bus": "0", "type": "tx", "frame": [{ "id": 2015, "dlc": 8, "rtr": false, "extd": false, "data": [2, 1, 91, 170, 170, 170, 170, 170] }] };return msg;
```

Receive: 
```
var obj = Object();

for (let i = 0; i < msg.payload.frame.length; i++) {
    if (msg.payload.frame[0].data[2] == 91) {
        obj.battery_pack_remaining = (Math.round((msg.payload.frame[0].data[3] * (1 / 2.55)) * 10) / 10);
        msg.payload = obj;
        return msg;
    }
}
```
![image](https://user-images.githubusercontent.com/20144319/224500294-c03933b5-1d22-468a-8d17-3e3fa8ff9a57.png)

**Comment by meatpiHQ** [comment]
@smurf12345 Good stuff!

Try sending this frame to initialize the Bus, while active charging and see if wakes up the ECU. Let us know if it works.

{"bus":"0","type":"tx","frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,0,170,170,170,170,170]}]}

**Reply by smurf12345** [reply]
@meatpiHQ  **That did the trick!  Thank you.**

It looks like once I send the frame to initialize the bus I have about 30 seconds (guesstimate) to send a request before it goes back to sleep.  Now, I got options to monitor the charge and potentially I can stop the charge at 50% and finish charging right before I leave based on calendar events or some other trigger, keeping the battery more within 50% range so it doesn't sit towards the higher end without use for days.

Awesome job on this device by they way!  I've been looking for something like this for years where I pull into the garage, it automatically connects, and pull data.
